src/components/Courses.js

```diff
@@ -1,149 +1,128 @@
   },
-
-  coursesDiv: {
+  loading: {
     display: 'flex',
-    flexFlow: 'wrap'
+    marginTop: 50,
+    justifyContent: 'center',
   },
-
-  coursesDivMobile: {
+  error: {
     display: 'flex',
-    flexDirection: 'column'
+    justifyContent: 'center',
   },
-
-  card: {
-    backgroundColor: '#fafafa'
+  container: {
+    marginTop: '1em',
   },
-
-  content: {
-    paddingTop: 0,
-    display: 'flex',
-    justifyContent: 'center'
+  actions: {
+    flexWrap: 'wrap',
+    justifyContent: 'center',
   },
-
-  empty: {
-    textAlign: 'center'
-  },
-
-  progress: {
-    margin: `0 ${theme.spacing(2)}px`
-  }
 }))
 
 export default function Courses(props) {
   const classes = useStyles()
-  const current_term = useSelector(state => state.terms.current_term)
-  const courses = useSelector(state => state.courses.courses)
-  const { mobile } = props
-  const dispatch = useDispatch();
-
-  useEffect(() => {
-    dispatch(fetch_courses(current_term))
-  }, [dispatch, current_term])
+  const courses = useSelector((state) => state.courses)
+  const courses_fetched = useSelector((state) => state.fetched)
+  const courses_fetching = useSelector((state) => state.fetching)
+  const courses_error = useSelector((state) => state.error)
 
-  if (courses === null || courses === []) {
+  if (courses_fetching === true) {
     return (
-      <Typography variant="h3" className={classes.empty} tabIndex="0">
+      <div className={classes.loading}>
+        <CircularProgress color='secondary' size={50} />
+      </div>
+    )
+  } else if (courses_fetched === true && courses_error === true) {
+    return (
+      <div className={classes.error}>
+        <ErrorMessages />
+      </div>
+    )
+  } else if (courses === null || courses === []) {
+    return (
+      <Typography variant='h3' className={classes.empty} tabIndex='0'>
         You currently have no courses for this semester.
       </Typography>
     )
-  }
-
-
-  let elements = []
-  for (let [i, course] of courses.entries()) {
-    if (course.meetings.length > 1 || course.instructors.length > 1) {
-      elements.push(
-        <ExpandableCourse
-          course={course}
-          key={'expandable' + Math.random()}
-          mobile={mobile}
-        />
-      )
-    } else if (!Object.is(course.waitList, '0')) {
-      elements.push(
-        <WaitlistCourse
-          course={course}
-          key={'waitlist' + Math.random()}
-          mobile={mobile}
-        />
-      )
-    } else {
-      elements.push(
-        <div
-          className={classes.courseContainer}
-          key={course.crn}
-        >
-          <div style={{ marginTop: '1em' }}>
-            <Card
-              className={classes.card}
-              key={course.crn + i + Math.random()}
-            >
-              <CourseHeader mobile={mobile} course={course} />
-              <CardContent
-                className={classes.content}
+  } else {
+    let elements = []
+    for (let [i, course] of courses.entries()) {
```

Comment:
Nice! I really like the `.entries()` - I actually wasn't aware of that method's existence 

Maybe eventually the different methods of iteration across the components can be made a bit more similar? This is a `for` loop on entries, other places do some explicit indexing based on sizes of arrays, other places do `foreach`

Reply:
i don't wanna burst your bubble, but i've gotten rid of pretty much all of the for loops in #212... lol. i switched to maps for everything

Reply:
Oh no that's fine. @kdjohnson can tell you - I was very against the use of `.map()` when I started at UTS, but I think they are my favorite sort of iteration now